src/ui/components/shared/APIKeys.tsx

```diff
@@ -0,0 +1,186 @@
+import React, { useEffect, useMemo, useState } from "react";
+import { ApiKey, ApiKeyResponse, ApiKeyScope } from "ui/types";
+
+import TextInput from "./Forms/TextInput";
+import MaterialIcon from "./MaterialIcon";
+
+const scopeLabels: Record<ApiKeyScope, string> = {
+  "create:recording": "Create recordings",
+  "write:sourcemap": "Upload source maps",
+};
+
+export default function APIKeys({
+  apiKeys,
+  description,
+  error,
+  loading,
+  addKey,
+  deleteKey,
+  scopes,
+}: {
+  apiKeys: ApiKey[];
+  scopes?: ApiKeyScope[];
+  description: string;
+  error: any;
+  loading: boolean;
+  addKey: (label: string, scopes?: string[]) => Promise<ApiKeyResponse>;
+  deleteKey: (id: string) => void;
+}) {
+  const [keyValue, setKeyValue] = useState<string>();
+  const [selectedScopes, selectScopes] = useState<ApiKeyScope[]>(scopes || []);
+  const [label, setLabel] = useState<string>("");
+  const [copied, setCopied] = useState(false);
+
+  useEffect(() => {
+    if (copied) {
+      setTimeout(() => setCopied(false), 2000);
+    }
+  }, [copied, setCopied]);
+
+  const canSubmit = label && !loading && selectedScopes.length > 0;
+  const sortedKeys = useMemo(
+    () =>
+      [...apiKeys].sort(
+        (a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
+      ),
+    [apiKeys]
+  );
+
+  return (
```

Comment:
More a style thing — can we refactor this jsx chunk for readability? Though I'm only 80% sure that it's more readable given the additional plumbing.

Reply:
Good call. The markup was getting a bit long. I refactored the list and "new key" code into separate components but left the meat of adding a key in the main component to avoid plumbing all the props about.